File: LongChallenge_2018_08_SHKNUM.cpp
// code_report Solution
// https://youtu.be/yMxoRT381yQ

#include <algorithm>
#include <iostream>
#include <cmath>
#include <set>

using namespace std;

auto precompute_m () {
   set<int> s;
   for (int i = 0; i <= 30; ++i)
      for (int j = 0; j <= 30; ++j)
         if (i != j) s.insert (pow (2, i) + pow (2, j));
   return s;
}

auto calculate_moves (const set<int>& s, int N) {
   if (N < 3) return 3 - N;     
   auto it = s.lower_bound (N);
   if (N == *it) return 0;      
   return min (N - *prev (it), *it - N);
}

int main () {
   auto s = precompute_m ();
   int t; cin >> t;
   while (t--) {
      int N; cin >> N;
      cout << calculate_moves (s, N) << endl;
   }
   return 0;
}  

File: LongChallenge_2018_08_SHKNUM.java
// code_report Solution
// https://youtu.be/yMxoRT381yQ

import java.util.*;
import java.io.*;

public class Main {

     public static void main(String []args)
     {
           TreeSet<Integer> s = new TreeSet<> ();

           for (int i = 0; i <= 30; ++i)
              for (int j = 0; j <= 30; ++j)
                 if (i != j) s.add ((int) (Math.pow (2, i) + Math.pow (2, j)));
                 
            Scanner sc = new Scanner (System.in);
            int     t  = sc.nextInt ();
            
            while (t-- != 0) {
                int N   = sc.nextInt ();
                int ans = 0;
                
                if (N < 3) ans = 3 - N;
                else       ans = Math.min (N - s.floor (N), s.higher (N) - N);
                
                System.out.println (ans);
            }
     }
}
